### PR TITLE
total variational training loss

### PR DESCRIPTION
**Pull request by 51N84D**
Adding total variational regularization for smoothing mask predictions (in the mask decoder)

### CODE REVIEW

**Comment by 51N84D** [comment]
I'm trying to allow multiple loss functions per update task. I'm sure there's a cleaner way to do it than this. What do you guys think?

File: omnigan/losses.py [hunk 1]
```diff
@@ -128,6 +128,25 @@ def __call__(self, S_real, S_fake):
         return self.cosine_loss(self.v_real_t, self.v_fake_t)
 
 
+class TVLoss(nn.Module):
```
Comment (vict0rsch) [inline omnigan/losses.py:131]: can you add a docstring? describing behavior + cite paper

File: omnigan/losses.py [hunk 2]
```diff
@@ -128,6 +128,25 @@ def __call__(self, S_real, S_fake):
...
+    def __init__(self, TVLoss_weight=1):
+        super(TVLoss, self).__init__()
+        self.TVLoss_weight = TVLoss_weight
```
Comment (vict0rsch) [inline omnigan/losses.py:134]: no initial capital letter if not class call

**Comment by vict0rsch** [comment]
> I'm trying to allow multiple loss functions per update task. I'm sure there's a cleaner way to do it than this. What do you guys think?

can you elaborate? I'm not sure I see the issue

File: omnigan/trainer.py [hunk 3]
```diff
@@ -940,6 +974,30 @@ def eval(self, num_threads=5, verbose=0):
...
+                            domain
+                        ] = update_loss.item()
+
+                        # Then TV loss
+                        update_loss = self.losses["G"]["tasks"][update_task]["tv"](
+                            prediction
+                        )
+                        self.logger.losses.task_loss[update_task]["tv"][
+                            domain
+                        ] = update_loss.item()
```
Comment (51N84D) [inline omnigan/trainer.py:1000]: @vict0rsch this is the part of the code I'm worried about. This could get messy if we have to treat each auxiliary task independently like this. Though it might be the case that this is only necessary for the mask decoder since that's our focal task and the other auxiliary tasks may not require additional losses.
